drivers: clock_control: nrf: Update driver to use nrfx_clock and nrfx_power

nordic-krch opened the pull request:
Updated driver implementation to use nrfx drivers.

nordic-krch commented:
@emob-nordic can you review usb part?

nordic-krch commented:
DNM until https://github.com/zephyrproject-rtos/hal_nordic/pull/45 is merged

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -499,38 +469,103 @@ void z_nrf_clock_control_lf_on(enum nrf_lfclk_start_mode start_mode)
+	nrfx_clock_enable();
+
+#if defined(CONFIG_USB_NRFX)
+	static const nrfx_power_usbevt_config_t config = {
```
This structure does not need to be `static`.

nordic-krch replied:
it can land in flash only if it is static. Compiling samples/subsys/cdc_acm with and without static show FLASH +12 after removing `static`.

anangl replied:
That's strange. `static` should not matter in this regard. As long as it is `const` it can be placed in flash, if that's suitable.

nordic-krch replied:
noticed it already in the past, seems that destination rule (static -> fixed address, no static ->stack) is evaluated before const keyword.

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -577,94 +613,13 @@ DEVICE_AND_API_INIT(clock_nrf, DT_INST_LABEL(0),
-		nrf_power_int_enable(NRF_POWER, mask);
-		irq_enable(DT_INST_IRQN(0));
+		nrfx_power_usbevt_enable();
 	} else {
-		nrf_power_int_disable(NRF_POWER, mask);
+		nrfx_power_usbevt_disable();
```
Why not entirely remove `nrf5_power_usb_power_int_enable()` (which is quite an artificial thing here) and call those functions directly from `usb_dc_nrfx.c`?

nordic-krch replied:
moved

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -499,38 +469,103 @@ void z_nrf_clock_control_lf_on(enum nrf_lfclk_start_mode start_mode)
+	}
+}
+
+#if defined(CONFIG_USB_NRFX)
```
Is there a reason why those USB related things are still kept here? It seems that they all could be moved to `usb_dc_nrfx.c`.

nordic-krch replied:
no reason, moved it to `usb_dc_nrfx.c`. Only left `USBREGULATOR_IRQn` interrupt connection because it is using same priority as clock power interrupt.

anangl replied:
I think it should be moved as well. It's purely USB related stuff. And you can get the power/clock IRQ priority also from `usb_dc_nrfx.c`, possibly with a comment why this priority is used.

nordic-krch replied:
done

emob-nordic commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -7,7 +7,7 @@
 #include "nrf_clock_calibration.h"
 #include <drivers/clock_control/nrf_clock_control.h>
-#include <hal/nrf_clock.h>
+#include <nrfx_clock.h>
```
Without `#include <nrfx_power.h>` samples for 5340 will not compile as `nrfx_power.h` includes `"nrfx_power_compat.h"` with usbreg stuff (for 5340).

emob-nordic commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -10,8 +10,8 @@
 #include "nrf_clock_calibration.h"
+#include <nrfx_clock.h>
 #include <logging/log.h>
-#include <hal/nrf_power.h>
```
Without `#include <nrfx_power.h>` samples for 5340 will not compile as `nrfx_power.h` includes `"nrfx_power_compat.h"` with usbreg stuff (for 5340).

nordic-krch replied:
moved all usb related stuff to `usb_dc_nrfx.c`

emob-nordic commented:
I think we can squash [this](https://github.com/emob-nordic/zephyr/commit/05e8a1ffd8a03b360ee52ca972eb9aed14a27ea7) to the 2nd commit. Ideally squash and later divide into 2: clock/usb related stuff.

nordic-krch committed:
drivers: clock_control: nrf: Make use of related nrfx drivers

Update the driver implementation to use nrfx_clock and nrfx_power
drivers. Update also revisions of the hal_nordic and nrf_hw_models
modules, so that it is actually possible to use those drivers.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>
Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

nordic-krch committed:
drivers: clock_control: nrf: Move USB part to the USB driver

Removed all nrf_power/USB related stuff from clock control
driver to USB driver.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

emob-nordic commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -23,6 +23,7 @@
 #include <hal/nrf_power.h>
```
This shall be removed. No longer needed.

nordic-krch replied:
done

emob-nordic commented on drivers/usb/device/usb_dc_nrfx.c:
```diff
@@ -493,22 +494,22 @@ static inline struct usbd_event *usbd_evt_alloc(void)
 }
 
-void usb_dc_nrfx_power_event_callback(nrf_power_event_t event)
+void usb_dc_power_event_handler(nrfx_power_usb_evt_t event)
```
`static` missing.

nordic-krch replied:
done

nordic-krch commented:
Seems like `nrf52_bsim` is a problem. nrfx so far was not compiled for nrf52_bsim and looks like there is a lot of issues...

@aescolar do you have any ideas how to handle that?

aescolar commented:
Hi @nordic-krch ,
Seems the kconfig dependencies are borked (this PR is also failing in qemu_cortex_m0).

> nrfx so far was not compiled for nrf52_bsim and looks like there is a lot of issues...

The nrfx is compiled in the nrf52_bsim (https://github.com/zephyrproject-rtos/nrf_hw_models/tree/master/src/nrfx) , but not the same version that you are compiling with the real HW. The version used with the HW models does not have the extra Kconfig options added explicitly in the nrfx module, so I presume that is why it is failing (and why it also fails in qemu)

nordic-krch commented:
> @anangl knows all about it :)

and he is enjoying his vacation time, good for him.

I have workarounded kconfig options for now but the problem I have seen was that once i include `hal_nordic` sources i will also get include paths and `<hal/nrf_clock.h>` includes version from nrfx and not from `nrf52_bsim`. I assumed that since hal is emulated in the models then i should be able to take nrfx_clock driver with `nrf52_bsim` hal.

aescolar commented:
> the problem I have seen was that once i include `hal_nordic` sources i will also get include paths and `<hal/nrf_clock.h>` includes version from nrfx and not from `nrf52_bsim`. 

Correct, you cannot compile with hal/nordic/nrfx/ when compiling for the nrf52_bsim, as both HALs versions will conflict.

> I assumed that since hal is emulated in the models then i should be able to take nrfx_clock driver with `nrf52_bsim` hal.

The drivers provided by the nrfx were not copied into the nrf52 HW models (only the needed HAL and MDK parts were) If you want to make use of some driver provided by the nrfx you'd need to get it into https://github.com/BabbleSim/ext_NRF52_hw_models/tree/master/src/nrfx/drivers

nordic-krch commented:
> The nrfx is compiled in the nrf52_bsim

@aescolar I am not sure that this is a good path to go. `nrfx_common` is in nrf52_bsim but this is just a bunch of helper macros. Once i tried to move nrfx driver (nrfx_clock.c) and its dependencies to nrf52_bsim it is like doing a complete copy of nrfx in nrf52_bsim.

I would expect that since nrf52_bsim is hw simulator then i can cut on hal level and use nrfx drivers but with hal from nrf52_bsim. Currently, the problem are header paths. I will see if there are other obstacles with that is workarounded (i dont yet have good fix for that).

aescolar commented:
> @aescolar I am not sure that this is a good path to go. `nrfx_common` is in nrf52_bsim but this is just a bunch of helper macros. Once i tried to move nrfx driver (nrfx_clock.c) and its dependencies to nrf52_bsim it is like doing a complete copy of nrfx in nrf52_bsim.

The nrfx in the nrf52_bsim repo is mostly a copy of the real one, with a few differences to account for sideeffects and for whatever is different in the models than in real HW. If the driver goes around the HAL in some cases you may also need to hack a bit it anyhow.
Anyhow note that the aim with that nrfx fork is to have all relevant HAL and MDK APIs there. So adding also some nrfx drivers/ is not really that bad.

nordic-krch commented:
@aescolar when i started to copy `nrfx_clock` and tried to run it i seen a lot of issues. I would need to make a copy of almost full nrfx infrastructure. I tried another path: modify repos to get hal from bsim but things above that (drivers, infrastructure) from nrfx. I ended where i have hello_world compiling for `nrf52_bsim` (with clock control using nrfx_clock). Here are the commits i've done:

hal_nordic:
- https://github.com/nordic-krch/hal_nordic/commit/d9a2e98e7b6f4d1de557b15a0302f92fa37debe1

bsim_hw_models:
- https://github.com/nordic-krch/ext_NRF52_hw_models/commit/7ae8736fb8a7bbfa019e878c0a33fbb557ef442b
- https://github.com/nordic-krch/ext_NRF52_hw_models/commit/19dece81681a9cefae4f2763fcf9216dc574e94e

zephyr:
- https://github.com/nordic-krch/zephyr/commit/eab3ed3c2a8e24f581937c84ce8dd838e4c55a65

They are fairly small and may allow direct use of nrfx drivers.

> If the driver goes around the HAL in some cases you may also need to hack a bit it anyhow.

Regarding that, i think that it should be nrfx goal to not do that (and most likely is). I've seen this in a workaround in nrfx_clock but was able to surpress that but turning off the workaround. I know that nrfx team is writing unit tests for drivers and mocks hal so it should not contain hardware dependent code in it's base code (may have it in workarounds which are ifdefed).

Anyway, @anangl is on vacation, i'm leaving for vacation in few days so we will get back to that in September.

carlescufi commented:
@nordic-krch could you please give permission to @anangl to pick this up once he comes back from vacation on the 24th? That way him and @aescolar can figure out a solution while you are on holiday.

nordic-krch commented:
@anangl feel free to pick this up once i'm on vacation.

aescolar commented:
> @aescolar when i started to copy `nrfx_clock` and tried to run it i seen a lot of issues. I would need to make a copy of almost full nrfx infrastructure.

That does not sound specially scary. But I'd want to ask @anangl about his plans here.

> I tried another path: modify repos to get hal from bsim but things above that (drivers, infrastructure) from nrfx. I ended where i have hello_world compiling for `nrf52_bsim` (with clock control using nrfx_clock)...

Getting hello_world running is easier than getting the BT stack running :). I would need to see how intermingled things end up, but I would fear too much.

emob-nordic commented:
@nordic-krch Could You please rebase onto recent master? There is a fix in USB i want to test nrf53 with.

carlescufi commented:
@kl-cruz this should include setting the `MCUSEL` register right?
```
NRF_P0->PIN_CNF[0] =         GPIO_PIN_CNF_MCUSEL_Peripheral << GPIO_PIN_CNF_MCUSEL_Pos;       NRF_P0->PIN_CNF[1] =         GPIO_PIN_CNF_MCUSEL_Peripheral << GPIO_PIN_CNF_MCUSEL_Pos;
```

kl-cruz commented:
> 
> 
> @kl-cruz this should include setting the `MCUSEL` register right?
> 
> ```
> NRF_P0->PIN_CNF[0] = GPIO_PIN_CNF_MCUSEL_Peripheral << GPIO_PIN_CNF_MCUSEL_Pos;
> NRF_P0->PIN_CNF[1] = GPIO_PIN_CNF_MCUSEL_Peripheral << GPIO_PIN_CNF_MCUSEL_Pos;
> ```

Such configuration must be done for nRF5340 on devkit 0.9.0. You can use nrf_gpio.h hal right now:
https://github.com/zephyrproject-rtos/hal_nordic/blob/master/nrfx/hal/nrf_gpio.h#L148

anangl commented:
> > @aescolar when i started to copy `nrfx_clock` and tried to run it i seen a lot of issues. I would need to make a copy of almost full nrfx infrastructure.
> 
> That does not sound specially scary. But I'd want to ask @anangl about his plans here.

@aescolar I would definitely like to avoid copying modified nrfx drivers to the nrf_hw_models module. This would make this module much harder to maintain. And as @nordic-krch already mentioned, it should be possible to use nrfx drivers in a simulated environment just by providing modified HALs, that was actually one of the reason we introduced those HALs. In case some driver turns out to be problematic in this regard, we'll need to improve that one.

> > I tried another path: modify repos to get hal from bsim but things above that (drivers, infrastructure) from nrfx. I ended where i have hello_world compiling for `nrf52_bsim` (with clock control using nrfx_clock)...
> 
> Getting hello_world running is easier than getting the BT stack running :). I would need to see how intermingled things end up, but I would fear too much.

I continued the path that @nordic-krch had started and I was able to successfully run the BT simulation with central_hr and peripheral samples. I'll promptly prepare a draft PR to show you the needed changes (and to check that approach in CI). Basically, these changes are even a bit simpler than those mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/27422#issuecomment-672755626.

anangl committed:
soc: nordic: nrf51: Add missing peripheral Kconfig options

To indicate that the CLOCK and POWER peripherals are present in those
SoCs, thus the corresponding nrfx drivers can be used.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

aescolar commented:
> @aescolar I would definitely like to avoid copying modified nrfx drivers to the nrf_hw_models module.

@anangl Fair enough. I guess we just need to have a clean way of selecting the drivers from the nrfx module, without the HAL & MDK. If we have that it should be easy enough for everybody.

anangl committed:
tests/bluetooth/ll_settings: Prevent switch-unreachable warning

After the fallthrough comments have been replaced (see commit
741ad17a37807a5396e73c10cf3e5fb425bc87bb) with a macro expanding
to the corresponding attribute, those hints became statements,
and if one happens to occur before any switch case, a warning
reporting it as an unreachable statement is generated.
Prevent the above from happening for function ll_rx_dequeue() in
ull.c by enabling an option that will make some switch case to
appear before this problematic statement.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Rebased on the latest master.

anangl commented:
Included the additions tested in #27776.

anangl commented:
The following two PRs must be merged first:
- https://github.com/BabbleSim/ext_NRF52_hw_models/pull/12
- https://github.com/zephyrproject-rtos/hal_nordic/pull/51

anangl commented:
Rebased on master.

anangl commented:
Updated revisions of `hal_nordic` and `nrf_hw_models` modules to proper SHAs.

nordic-krch commented:
it would be worth renaming now `nrf_power_clock.c` to `clock_control_nrfx.c`. What do you think @anangl ?

nordic-krch committed:
drivers: clock_control: Rename nrf_power_clock to clock_control_nrf

Renamed nrf clock control driver file since it does no long control
POWER peripheral.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@carlescufi 
> @kl-cruz this should include setting the MCUSEL register right?

Regarding this nrf53 part - i would prefer to do that in another PR since it does not seem related.
